packages/next/src/server/app-render.tsx

```diff
@@ -925,6 +925,10 @@ export async function renderToHTMLOrFlight(
     typeof actionId === 'string' &&
     req.method === 'POST'
 
+  const protocol = req.headers['x-forwarded-proto'] || 'http'
+  const host = req.headers.host
```

Comment:
This is potentially dangerous when this incoming request header is trusted in an environment where it is passed through directly. Can you have a look at how this is handled for `edge` where a full URL is required as well and mirror that approach?